File: Winter2022/lab-solutions/java/Lab9/SensorLabSkeleton/app/src/main/java/course/labs/graphicslab/BubbleActivity.java
package course.labs.graphicslab;

import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Insets;
import android.graphics.Paint;
import android.graphics.Point;
import android.graphics.Rect;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.Display;
import android.view.GestureDetector;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowInsets;
import android.view.WindowMetrics;
import android.widget.RelativeLayout;
import android.widget.TextView;
import android.widget.Toast;

import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

public class BubbleActivity extends Activity implements SensorEventListener {

	private static final String TAG = "BubbleActivity";

    private static final int NO_FILTER = 0;
    private static final int LOW_PASS_FILTER = 1;
    private static final int HIGH_PASS_FILTER = 2;
    private static int mFilter = NO_FILTER;

	// The Main view
	private RelativeLayout mFrame;

	// Bubble image's bitmap
	private Bitmap mBitmap;

	// Display dimensions
	private int mDisplayWidth, mDisplayHeight;

    // Gesture Detector
    private GestureDetector mGestureDetector;

    // A TextView to hold the player message
    private TextView mPlayerMessage;

    // Attribute to storing the last sensor update
    private long lastUpdate;

    // Arrays for storing filtered values
    private final float[] gravity = new float[3];
    private final float[] acceleration = new float[3];

    // TODO
    //  Create members for a SensorManager and Sensor
    private SensorManager sensorManager;
    private Sensor accelerometer;

    @Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		setContentView(R.layout.main);

		// Set up user interface
		mFrame = findViewById(R.id.frame);

        // Set up text view
        mPlayerMessage = findViewById(R.id.startMessage);

		// Load basic bubble Bitmap
		mBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.ball);

        // TODO
        //  Get a SensorManager, and then use it to get an accelerometer Sensor.
        //  Set the SensorManager and Sensor members that you defined above appropriately.
        //  If the device does not have an accelerometer, display a message indicating so and exit.
        sensorManager = (SensorManager) getSystemService(SENSOR_SERVICE);

        if (null != sensorManager) {
            accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);

            if (null == accelerometer) {
                Toast.makeText(getApplicationContext(), getString(R.string.accelerometer_error), Toast.LENGTH_SHORT).show();
                finish();
            }
        }

        // Calculate display size
        Pair<Integer, Integer> dimensions = getScreenDimensions(this);

        // Subtract diameter of the ball from width and height
        mDisplayWidth = dimensions.getFirst() - BubbleView.SCALED_BITMAP_SIZE;
        mDisplayHeight = dimensions.getSecond() - BubbleView.SCALED_BITMAP_SIZE;
	}

    private Pair<Integer, Integer> getScreenDimensions(Activity activity) {
        int width;
        int height;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            final WindowMetrics metrics = getWindowManager().getCurrentWindowMetrics();
            final WindowInsets windowInsets = metrics.getWindowInsets();
            Insets insets = windowInsets.getInsetsIgnoringVisibility(WindowInsets.Type.navigationBars()
                    | WindowInsets.Type.displayCutout());

            int insetsWidth = insets.right + insets.left;
            int insetsHeight = insets.top + insets.bottom;

            final Rect bounds = metrics.getBounds();
            width = bounds.width() - insetsWidth;
            height = bounds.height() - insetsHeight;
        } else {
            Display display = activity.getWindowManager().getDefaultDisplay();
            Point size = new Point();
            display.getSize(size);
            width = size.x;
            height = size.y;
        }

        return new Pair<>(width, height);
    }

	@Override
	protected void onResume() {
		super.onResume();

        // TODO
        //  Register a listener for the accelerometer
        sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_UI);
        lastUpdate = System.currentTimeMillis();

        setupGestureDetector();

    }

    public void onSensorChanged(SensorEvent event) {
        // TODO
        //  Apply a low- and high-pass filter to the raw sensor values
        //  HINT: See https://developer.android.com/guide/topics/sensors/sensors_motion#sensors-motion-accel
        //        for more information
        if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
            long actualTime = System.currentTimeMillis();

            if (actualTime - lastUpdate > BubbleView.REFRESH_RATE) {
                lastUpdate = actualTime;
                float x = event.values[0];
                float y = event.values[1];
                float z = event.values[2];

                // Isolate the force of gravity with the low-pass filter.
                gravity[0] = applyLowPassFilter(x, gravity[0]);
                gravity[1] = applyLowPassFilter(y, gravity[1]);
                gravity[2] = applyLowPassFilter(z, gravity[2]);

                // Remove the gravity contribution with the high-pass filter.
                acceleration[0] = applyHighPassFilter(event.values[0], gravity[0]);
                acceleration[1] = applyHighPassFilter(event.values[1], gravity[1]);
                acceleration[2] = applyHighPassFilter(event.values[2], gravity[2]);

                // TODO
                //  If there is a BubbleView, use its setSpeedAndDirection() method
                //  to set its speed and direction based on the sensor values and the
                //  current setting of mFilter, which will be one of NO_FILTER, LOW_PASS_FILTER, or HIGH_PASS_FILTER.
                BubbleView bubbleView = (BubbleView) mFrame.getChildAt(0);
                if (null != bubbleView) {
                    if (mFilter == NO_FILTER) {
                        bubbleView.setSpeedAndDirection(x, y);
                    } else if (mFilter == LOW_PASS_FILTER) {
                        bubbleView.setSpeedAndDirection(gravity[0], gravity[1]);
                    } else if (mFilter == HIGH_PASS_FILTER) {
                        bubbleView.setSpeedAndDirection(acceleration[0], acceleration[1]);
                    } else {
                        Log.w(TAG, "Unknown filter type " + mFilter + ".  No action taken");
                    }
                }
            }
        }
    }

    // De-emphasize transient forces
    private float applyLowPassFilter(float current, float gravity) {
        final float ALPHA = 0.8f;
        return gravity * ALPHA + current * (1 - ALPHA);
    }

    // De-emphasize constant forces
    private float applyHighPassFilter(float current, float gravity) {
        return current - gravity;
    }

    // Nothing to do here, just note that onAccuracyChanged must be implemented
    @Override
    public final void onAccuracyChanged(Sensor sensor, int accuracy) {
        // Do something here if sensor accuracy changes.
        // For now we will leave this unimplemented.
    }

	// Set up GestureDetector
	private void setupGestureDetector() {
		mGestureDetector = new GestureDetector(this, new GestureDetector.SimpleOnGestureListener() {
            // If there is a BubbleView, and a single tap intersects it, remove it.
            // If there are no BubbleViews, create a new BubbleView at the tap's location
            // and add it to mFrame.
			@Override
			public boolean onSingleTapConfirmed(MotionEvent event) {
                float x = event.getRawX();
                float y = event.getRawY();

                int numBubbles = mFrame.getChildCount();

                if (numBubbles != 0) {
                    BubbleView bv = (BubbleView) mFrame.getChildAt(0);
                    if (bv.intersects(x, y)) {
                        bv.stopMovement();
                     }
                 } else {
                    Context context = mFrame.getContext();
                    BubbleView bubbleView = new BubbleView(context, x, y);
                    mFrame.addView(bubbleView);
                    bubbleView.startMovement();
                    setPlayerMessage();
                }
				return true;
			}

            @Override
            public void onLongPress(MotionEvent e) {
                // Cycle through the filter options
                mFilter = (mFilter + 1) % 3;
                setPlayerMessage();
            }
        });
	}

    // Update the message that is displayed to the player
    public void setPlayerMessage() {
        String message;
        if (mFrame.getChildCount() == 0) {
            message = getString(R.string.start_message);
        }
        else {
            if (mFilter == NO_FILTER) {
                message = getString(R.string.no_filter_message);
            } else if (mFilter == LOW_PASS_FILTER) {
                message = getString(R.string.lowpass_filter_message);
            } else {
                message = getString(R.string.highpass_filter_message);
            }
        }
        mPlayerMessage.setText(message);
    }

	@Override
	public boolean onTouchEvent(MotionEvent event) {
        return mGestureDetector.onTouchEvent(event);
	}

	@Override
	protected void onPause() {
		super.onPause();

        // TODO
        //  Unregister the sensor event listener
        sensorManager.unregisterListener(this);
	}

	// BubbleView is a View that displays a bubble.
	// This class handles animating, drawing, and popping amongst other actions.
	public class BubbleView extends View {
		private static final int BITMAP_SIZE = 64;
		private static final int REFRESH_RATE = 5;
		private final Paint mPainter = new Paint();
		private ScheduledFuture<?> mMoverFuture;
		private static final int SCALED_BITMAP_SIZE = BITMAP_SIZE* 2;
		private final Bitmap mScaledBitmap;

		// location and direction of the bubble
		private float mXPos;
        private float mYPos;
        private final float mRadius;

        // Speed of bubble
        public float mDx, mDy;

        // Rotation and speed of rotation of the bubble
        private long mRotate;
        private final long mDRotate;

		BubbleView(Context context, float x, float y) {
			super(context);

			// Creates the bubble bitmap for this BubbleView
            mScaledBitmap = Bitmap.createScaledBitmap(mBitmap, SCALED_BITMAP_SIZE,
                    SCALED_BITMAP_SIZE, true);

			// Radius of the Bitmap
			mRadius = SCALED_BITMAP_SIZE / 2.0f;

			// Adjust position to center the bubble under user's finger
			mXPos = x - mRadius;
			mYPos = y - mRadius;

            // Set speed to 0 initially; it will be updated soon based on sensor input
            mDx = 0;
            mDy = 0;

            // Set speed of rotation to 5
            mDRotate = 5;

			mPainter.setAntiAlias(true);

		}

        // setSpeedAndDirection called by onSensorChanged(), values based on
        // accelerometer data and scaled
        private void setSpeedAndDirection(float x, float y) {
		    // NOTE the layout is set to landscape.  This means that the coordinate
            // system does not change, however, the range for x and y are essentially reversed
            mDx = y;
            mDy = x;

            // TODO
            //  Once your app is working, experiment with alternative
            //  ways of calculating mDx and mDy based on x and y to change
            //  the "feel". Some ideas are shown below.

            // Example 1: Uncomment this to make the ball go faster!
            // mDx = 2 * y;
            // mDy = 2 * x;

            // Example 2: Uncomment this to make the ball accelerate based on sensor
            // input. You can also scale the contribution of x and y.
            mDx += y;
            mDy += x;
        }

		// Start moving the BubbleView & updating the display
		private void startMovement() {
			// Creates a WorkerThread
			ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);

			// Execute the run() in Worker Thread every REFRESH_RATE
			// milliseconds
			// Save reference to this job in mMoverFuture
			mMoverFuture = executor.scheduleWithFixedDelay(() -> {
                doMove();
                BubbleView.this.postInvalidate();
            }, 0, REFRESH_RATE, TimeUnit.MILLISECONDS);
		}

        private void stopMovement() {
            if (null != mMoverFuture) {

                if (!mMoverFuture.isDone()) {
                    mMoverFuture.cancel(true);
                }

                // This work will be performed on the UI Thread
                mFrame.post(() -> {
                    mFrame.removeView(BubbleView.this);
                    setPlayerMessage();
                });
            }
        }

        // Return true if x and y intersect the position of the Bubble
		private synchronized boolean intersects(float x, float y) {
            float centerX = mXPos + mRadius;
            float centerY = mYPos + mRadius;
            return Math.sqrt(Math.pow(centerX - x, 2) + Math.pow(centerY - y, 2)) <= mRadius;
		}

		// Draw the Bubble at its current location
		@Override
		protected synchronized void onDraw(Canvas canvas) {
            canvas.save();
            mRotate += mDRotate;
            canvas.rotate(mRotate, mXPos + mRadius, mYPos + mRadius);
            canvas.drawBitmap(mScaledBitmap, mXPos, mYPos, mPainter);
            canvas.restore();
		}

        // Move the Bubble
        private synchronized void doMove() {
            // Don't let the bubble go beyond the edge of the screen
            // Set the speed to 0 if the bubble hits an edge.
            mXPos += mDx;
            if (mXPos >= mDisplayWidth) {
                mXPos = mDisplayWidth;
                mDx = 0;
            } else if (mXPos <= 0) {
                mXPos = 0;
                mDx = 0;
            }

            mYPos += mDy;
            if (mYPos >= mDisplayHeight) {
                mYPos = mDisplayHeight;
                mDy = 0;
            } else if (mYPos <= 0) {
                mYPos = 0;
                mDy = 0;
            }
        }
	}
}

File: Fall2024/labs/Lab7/BubbleGameSkeleton/app/src/main/java/mobiledev/unb/ca/graphicslab/BubbleActivity.kt
package mobiledev.unb.ca.graphicslab

import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Point
import android.media.AudioManager
import android.media.SoundPool
import android.os.Build
import android.os.Bundle
import android.view.GestureDetector
import android.view.GestureDetector.SimpleOnGestureListener
import android.view.MotionEvent
import android.view.View
import android.view.WindowInsets
import android.widget.RelativeLayout
import android.widget.TextView
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.Random

class BubbleActivity : Activity() {
    // The Main view
    private var mFrame: RelativeLayout? = null

    // Bubble image's bitmap
    private var bitmap: Bitmap? = null

    // Display dimensions
    private var displayWidth = 0
    private var displayHeight = 0

    // Gesture Detector
    private var gestureDetector: GestureDetector? = null

    // A TextView to hold the current number of bubbles
    private var bubbleCountTextView: TextView? = null

    // Sound variables
    // SoundPool
    private var soundPool: SoundPool? = null

    // ID for the bubble popping sound
    private var soundID = 0

    // Audio volume
    private var streamVolume = 0f

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main)

        // Set up user interface
        mFrame = findViewById(R.id.frame)
        bubbleCountTextView = findViewById(R.id.bubbles_text)

        // Initialize the number of bubbles
        updateNumBubblesTextView()

        // Load basic bubble Bitmap
        bitmap = BitmapFactory.decodeResource(resources, R.drawable.b64)
    }

    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            // Determine the screen size
            val (width, height) = getScreenDimensions(this)
            displayWidth = width
            displayHeight = height
        }
    }
    override fun onResume() {
        super.onResume()
        setStreamVolume()

        // TODO
        //  Make a new SoundPool, allowing up to 10 streams
        //  Store this as soundPool

        // TODO
        //  Set a SoundPool OnLoadCompletedListener that calls setupGestureDetector()
        soundPool!!.setOnLoadCompleteListener { _: SoundPool?, _: Int, status: Int ->
            // TODO Call setupGestureListener
        }

        // TODO
        //  Load the sound from res/raw/bubble_pop.wav
        //  Store this as soundID
    }

    // Method used to update the text view with the number of in view bubbles
    private fun updateNumBubblesTextView() {
        val text = getString(R.string.txt_number_of_bubbles, mFrame!!.childCount)
        bubbleCountTextView!!.text = text
    }

    // Retrieves the screen dimensions
    private fun getScreenDimensions(activity: Activity): Pair<Int, Int> {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            val windowMetrics = activity.windowManager.currentWindowMetrics
            val windowInsets: WindowInsets = windowMetrics.windowInsets
            val insets = windowInsets.getInsetsIgnoringVisibility(
                WindowInsets.Type.navigationBars() or WindowInsets.Type.displayCutout())

            val insetsWidth = insets.right + insets.left
            val insetsHeight = insets.top + insets.bottom

            val b = windowMetrics.bounds
            Pair(b.width() - insetsWidth, b.height() - insetsHeight)
        } else {
            val size = Point()
            @Suppress("DEPRECATION")
            val display = activity.windowManager.defaultDisplay // deprecated in API 30
            @Suppress("DEPRECATION")
            display?.getSize(size) // deprecated in API 30
            Pair(size.x, size.y)
        }
    }

    // Setup the stream volume
    private fun setStreamVolume() {
        // AudioManager audio settings for adjusting the volume
        val audioManager = getSystemService(AUDIO_SERVICE) as AudioManager

        // Get the current volume Index of particular stream type
        // Stream type is set to use AudioManager.STREAM_MUSIC
        val currentVolumeIndex = audioManager.getStreamVolume(STREAM_TYPE)
            .toFloat()

        // Get the maximum volume index for a particular stream type
        val maxVolumeIndex = audioManager.getStreamMaxVolume(STREAM_TYPE)
            .toFloat()

        // Set the volume between 0 --> 1
        streamVolume = currentVolumeIndex / maxVolumeIndex
        volumeControlStream = STREAM_TYPE
    }

    // Set up GestureDetector
    private fun setupGestureDetector() {
        gestureDetector = GestureDetector(this,
            object : SimpleOnGestureListener() {
                override fun onDown(motionEvent: MotionEvent): Boolean {
                    return true
                }

                // If a fling gesture starts on a BubbleView then change the
                // BubbleView's velocity based on x and y velocity from
                // this gesture
                override fun onFling(
                    event1: MotionEvent?, event2: MotionEvent,
                    velocityX: Float, velocityY: Float,
                ): Boolean {
                    // TODO
                    //  Implement onFling actions (See comment above for expected behaviour)
                    //  You can get all Views in mFrame one at a time using the ViewGroup.getChildAt() method

                    return true
                }

                // If a single tap intersects a BubbleView, then pop the BubbleView
                // Otherwise, create a new BubbleView at the tap's location and add
                // it to mFrame. Hint: Don't forget to start the movement of the
                // BubbleView.
                // Also update the number of bubbles displayed in the appropriate TextView
                override fun onSingleTapConfirmed(event: MotionEvent): Boolean {
                    // TODO - Implement onSingleTapConfirmed actions.
                    //  (See comment above for expected behaviour.)
                    //  You can get all Views in mFrame using the
                    //  ViewGroup.getChildCount() method

                    return true
                }
            })
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        // TODO
        //  Delegate the touch to the gestureDetector

        // Remove this when you're done the above TODO
        return true
    }

    override fun onPause() {
        super.onPause()

        // TODO
        //  Release all SoundPool resources
    }

    // BubbleView is a View that displays a bubble.
    // This class handles animating, drawing, and popping amongst other actions.
    // A new BubbleView is created for each bubble on the display
    inner class BubbleView internal constructor(context: Context?, x: Float, y: Float) :
        View(context) {
        private val mPainter = Paint()
        private var mMoverFuture: ScheduledFuture<*>? = null
        private var scaledBitmapSize = 0
        private var scaledBitmap: Bitmap? = null

        // location and direction of the bubble
        private var xPos: Float
        private var yPos: Float
        private val radius: Float

        // Speed of bubble
        private var mDx = 0f
        private var mDy = 0f

        // Rotation and speed of rotation of the bubble
        private var mRotate: Long = 0
        private var mDRotate: Long = 0

        private fun setRotation(r: Random) {
            // TODO
            //  Set rotation in range [1..5]
        }

        private fun setSpeedAndDirection(r: Random) {
            // TODO
            //  Set mDx and mDy to indicate movement direction and speed
            //  Limit speed in the x and y direction to [-3..3] pixels per movement
        }

        private fun createScaledBitmap(r: Random) {
            // TODO
            //  Set scaled bitmap size (scaledBitmapSize) in range [2..4] * BITMAP_SIZE


            // TODO
            //  Create the scaled bitmap (scaledBitmap) using size set above
        }

        // Start moving the BubbleView & updating the display
        fun startMovement() {
            // Creates a WorkerThread
            val executor = Executors.newScheduledThreadPool(1)

            // Execute the run() in Worker Thread every REFRESH_RATE
            // milliseconds
            // Save reference to this job in mMoverFuture
            mMoverFuture = executor.scheduleWithFixedDelay({
                // TODO
                //  Implement movement logic.
                //  Each time this method is run the BubbleView should
                //  move one step. (Use moveWhileOnScreen() to do this.)
                //  If the BubbleView exits the display, stop the BubbleView's
                //  Worker Thread. (Use stopMovement() to do this.) Otherwise,
                //  request that the BubbleView be redrawn.
            }, 0, Companion.REFRESH_RATE.toLong(), TimeUnit.MILLISECONDS)
        }

        // Returns true if the BubbleView intersects position (x,y)
        @Synchronized
        fun intersects(x: Float, y: Float): Boolean {
            val centerX = xPos + radius
            val centerY = yPos + radius

            // TODO
            //  Return true if the BubbleView intersects position (x,y)

            // Remove this when you're done the above TODO
            return false
        }

        // Cancel the Bubble's movement
        // Remove Bubble from mFrame
        // Play pop sound if the BubbleView was popped
        fun stopMovement(wasPopped: Boolean) {
            if (null != mMoverFuture) {
                if (!mMoverFuture!!.isDone) {
                    mMoverFuture!!.cancel(true)
                }

                // This work will be performed on the UI Thread
                mFrame!!.post {
                    // TODO
                    //  Remove the BubbleView from mFrame

                    // TODO
                    //  Update the TextView displaying the number of bubbles

                    // TODO
                    //  If the bubble was popped by user play the popping sound
                    //  HINT: Use the streamVolume for left and right volume parameters
                }
            }
        }

        // Change the Bubble's speed and direction
        @Synchronized
        fun deflect(velocityX: Float, velocityY: Float) {
            mDx = velocityX / Companion.REFRESH_RATE
            mDy = velocityY / Companion.REFRESH_RATE
        }

        // Draw the Bubble at its current location
        @Synchronized
        override fun onDraw(canvas: Canvas) {
            // TODO
            //  Save the canvas

            // TODO
            //  Increase the rotation of the original image by mDRotate

            // TODO
            //  Rotate the canvas by current rotation
            //  Hint - Rotate around the bubble's center, not its position

            // TODO
            //  Draw the bitmap at it's new location

            // TODO
            //  Restore the canvas
        }

        // Returns true if the BubbleView is still on the screen after the move
        // operation
        @Synchronized
        private fun moveWhileOnScreen(): Boolean {
            // TODO
            //  Move the BubbleView

            return isInView()
        }

        private fun isInView(): Boolean {
            // TODO
            //  Return true if the BubbleView is still on the screen after
            //  the move operation

            // Remove this when you're done the above TODO
            return false
        }

        init {
            // Create a new random number generator to
            // randomize size, rotation, speed and direction
            val r = Random()

            // Creates the bubble bitmap for this BubbleView
            createScaledBitmap(r)

            // Radius of the Bitmap
            radius = (scaledBitmapSize / 2).toFloat()

            // Adjust position to center the bubble under user's finger
            xPos = x - radius
            yPos = y - radius

            // Set the BubbleView's speed and direction
            setSpeedAndDirection(r)

            // Set the BubbleView's rotation
            setRotation(r)
            mPainter.isAntiAlias = true
        }
    }

    companion object {
        private const val TAG = "BubbleActivity"
        private const val STREAM_TYPE = AudioManager.STREAM_MUSIC
        private const val SOUND_POOL_MAX_STREAMS = 10
        private const val BITMAP_SIZE = 64
        private const val REFRESH_RATE = 40
    }
}